### PR TITLE
Ruby code sample for jobs

### PR DESCRIPTION
**Pull request by baochanzheng**
Files under jobs/V3 sample usages for the APIs jobs team provides.
jobs/spec/samples_spec.rb is the test file, verifying the sample files under jobs/V3.

### CODE REVIEW

**Comment by baochanzheng** [comment]
Hi Frank, have you gotten a chance to review this? @frankyn

File: jobs/V3/auto_complete_sample.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+require_relative 'basic_company_sample'
+require_relative 'basic_job_sample'
+
+=begin
+	The sample in this file introduce how to do a histogram search.
+=end
+
+class AutoCompleteSample
+	# Instantiate the client
+	@@Jobs   = Google::Apis::JobsV3
```
Comment (frankyn) [inline jobs/V3/auto_complete_sample.rb]: Don't use class variables.

https://github.com/github/rubocop-github/blob/master/STYLEGUIDE.md

File: jobs/V3/auto_complete_sample.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+# [END default_auto_complete]
+
+end
+
+
+# Test main. Run only if file is being executed directly or being called by ../spec/samples_spec.rb
+if (ARGV.include? File.basename(__FILE__)) || 
+	((File.basename(caller[0]).include? "samples_spec.rb") && (File.basename(caller[0]).include? "load"))
+	# test
+	company = BasicCompanySample.new
```
Comment (frankyn) [inline jobs/V3/auto_complete_sample.rb]: Keep this section simple. 

For example: 
https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/2ef920853952b386e644be14bebec3e2471b7edf/dlp/dlp_inspect_image_file.rb#L63

File: jobs/V3/basic_company_sample.rb [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		end
+		
+	end
+# [END create_company]
+
+# [START get_company]
+=begin 
+		Get a company
+=end
+	def getCompany(companyName)
```
[the remark this answers is not in the document]
Comment (baochanzheng) [reply jobs/V3/basic_company_sample.rb]: Hi Frank, thanks for the review! Do you mean each file can only have one test? I don't quite understand the reason... Plus, do I also need to implement command interface like some of the other ruby code samples did? I'd really love to have a GVC to discuss more details. I sent an invitation to you based on our calendars. Pls take or reschedule it if needed. :) 
Regards, 
Baochan

[the remark this answers is not in the document]
Comment (frankyn) [reply jobs/V3/basic_company_sample.rb]: Sg, thanks for setting it up!

**Comment by baochanzheng** [comment]
@frankyn Hi Frank, I made another commit, refactoring the basic_company_sample.rb and basic_job_sample.rb. Could you pls take a look at those two files and provide feedback? If this refactor looks good, I will apply it to all the other files as well. Thanks and regards, Baochan :)

File: jobs/V3/basic_job_sample.rb [hunk 4]
```diff
@@ -12,213 +12,228 @@
...
-			elsif getJob(jobName).nil?
-				puts "job doesn't exist";
-			end
-		end
+	# [END delete_job]
+end
+
+def run_basic_job_sample arguments
+
+	require_relative "basic_company_sample"
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: this line isn't required.

File: jobs/V3/basic_job_sample.rb [hunk 5]
```diff
@@ -12,213 +12,228 @@
...
+
+	command = arguments.shift
+	default_project_id = "projects/" + ENV["GOOGLE_CLOUD_PROJECT"];
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: Use interpolation:
"project/#{ENV["GOOGLE_CLOUD_PROJECT"]}"

File: jobs/V3/basic_job_sample.rb [hunk 6]
```diff
@@ -12,213 +12,228 @@
...
+	return job_generated;
+	# [END generate basic job]
+end
+
+def create_job(job_to_be_created)
+	# [START create_job]
+	require "google/apis/jobs_v3"
+
+	jobs   = Google::Apis::JobsV3;
+	default_project_id = "projects/" + ENV["GOOGLE_CLOUD_PROJECT"];
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: Should be passed in by a parameter.

File: jobs/V3/basic_job_sample.rb [hunk 7]
```diff
@@ -12,213 +12,228 @@
...
+
+	job_generated.compensation_info = compensation_info;
+	puts "Job generated: " + job_generated.to_json;
+	return job_generated;
+	# [END generate basic job]
+end
+
+def create_job(job_to_be_created)
+	# [START create_job]
+	require "google/apis/jobs_v3"
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: Add comments about what each parameter is w.r.t to https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/master/CONTRIBUTING.md#reference-snippets

File: jobs/V3/basic_job_sample.rb [hunk 8]
```diff
@@ -12,213 +12,228 @@
...
+	job_generated.company_name = company_name;
+	job_generated.employment_types = Array["FULL_TIME"];
+	job_generated.language_code = "en-US";
+	job_generated.application_info = application_info;
+	job_generated.description = "Design, develop, test, deploy, maintain and improve software.";
+	
+	# set compensation to 12 USD/hour
+	compensation_info = jobs::CompensationInfo.new;
+	compensation_entry = jobs::CompensationEntry.new;
+	compensation_amount = jobs::Money.new;
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: Could you try setting the value for compensation_amount using the following?:
```ruby
{
   currency_code: "USD",
   units: 12,
}
```

If it works for you apply it to other object representations. It's hard to parse each instance representing parts of the request.

File: jobs/V3/basic_company_sample.rb [hunk 9]
```diff
@@ -12,192 +12,197 @@
...
-			return companyCreated;
-		rescue
-			puts "Got exception while creating company"
-		end
-		
+	begin
+		create_company_request = jobs::CreateCompanyRequest.new;
+		create_company_request.company = company_to_be_created;
+		company_created = talentSolution_client.create_company(default_project_id, create_company_request);
+		puts "Company created: " + company_created.to_json;
```
Comment (frankyn) [inline jobs/V3/basic_company_sample.rb]: Use "Company created: #{company_created.to_json}"

File: jobs/V3/basic_job_sample.rb (hunk 4 above)
Comment (baochanzheng) [reply jobs/V3/basic_job_sample.rb]: I think it is required... I am using the functions in basic_company_sample to generate the company before creating jobs. If I didn't have this line, I can not refer to that function. Pls correct me if I was wrong.

Comment (frankyn) [reply jobs/V3/basic_job_sample.rb]: Ah sorry my mistake. You're correct that you need this to do that.

Why are you generating the company before creating the jobs? I would expect a CLI to create companies and another separate CLI to create jobs based on the created companies. At that point it becomes more of an end-user task to provide the information.

Comment (baochanzheng) [reply jobs/V3/basic_job_sample.rb]: Make sense. What about this: I take company_name as user input when they create job, and if the company exists, I create job under that company; if not I create the company for them before creating job? What do you think?

Comment (frankyn) [reply jobs/V3/basic_job_sample.rb]: The samples should focus on what the API should demonstrate to a developer:
Task one: Create a new company. **Error checking** to make sure the company doesn't exist yet is okay here.
Task two: Create a new job with a company name that should exist. **Error checking** to make sure the Company exists is fine but do not create a new company in this sample.

File: jobs/V3/basic_job_sample.rb (hunk 8 above)
Comment (baochanzheng) [reply jobs/V3/basic_job_sample.rb]: I made the changes based on my understand. Pls check whether it's as expected. Thanks! :)

File: jobs/spec/samples_spec.rb [hunk 10]
```diff
@@ -67,36 +75,41 @@
...
+                                               job_to_be_updated: job_created_test
+      job_discovery_delete_job job_name: job_created_test.name
+      job_discovery_delete_company company_name: company_created_test.name
       capture = $stdout.string
-      expect(capture).to include("Job created");
-      expect(capture).to include("Job got");
-      expect(capture).to include("Job updated");
-      expect(capture).to include("Job updated with filedMask title");
-      expect(capture).to include("Job deleted");
+      expect(capture).to include("Job created")
```
Comment (frankyn) [inline jobs/spec/samples_spec.rb]: Could you verify each test action was successful by requesting state from the Jobs service?

File: jobs/spec/samples_spec.rb [hunk 11]
```diff
@@ -67,36 +75,41 @@
...
+      expect(capture).to include("Job got")
+      expect(capture).to include("Job updated")
+      expect(capture).to include("Job updated with filedMask title")
+      expect(capture).to include("Job deleted")
     rescue
```
Comment (frankyn) [inline jobs/spec/samples_spec.rb]: Capture the rescue error and print it similar as in samples. It will help with future debugging.

File: jobs/V3/basic_job_sample.rb [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # Copyright 2018 Google, Inc
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache License, Version 2.0 (the "License")
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: Semicolon is expected here.

File: jobs/V3/basic_company_sample.rb [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # Copyright 2018 Google, Inc
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache License, Version 2.0 (the "License")
```
Comment (frankyn) [inline jobs/V3/basic_company_sample.rb]: Semicolon is expected here.
https://www.apache.org/licenses/LICENSE-2.0

File: jobs/V3/basic_job_sample.rb [hunk 14]
```diff
@@ -12,171 +12,172 @@
...
-	compensation_info.entries = Array[compensation_entry];
-
-	job_generated.compensation_info = compensation_info;
-	puts "Job generated: " + job_generated.to_json;
-	return job_generated;
+	compensation_entry = jobs::CompensationEntry.new :type => "BASE",
+													 :unit => "HOURLY",
+													 :amount => (jobs::Money.new :currency_code => "USD",
+										  										 :units => 12)
+	compensation_info = jobs::CompensationInfo.new :entries => Array[compensation_entry]
```
Comment (frankyn) [inline jobs/V3/basic_job_sample.rb]: An `Array` can be define with only `[]`.

File: jobs/V3/basic_company_sample.rb [hunk 15]
```diff
@@ -12,191 +12,205 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-def generate_company
+def job_discovery_generate_company display_name:, headquarters_address:
```
[the remark this answers is not in the document]
Comment (baochanzheng) [reply jobs/V3/basic_company_sample.rb]: I'm confused... 
1. Do I need to add job_discovery_ as prefix to all the functions?
2. Do all the test functions have to call service? I can't have it as separate method?

File: jobs/V3/basic_company_sample.rb [hunk 16]
```diff
@@ -12,191 +12,205 @@
...
-		return company_updated;
+		update_company_request = jobs::UpdateCompanyRequest.new
+		update_company_request.company = company_updated
+		company_updated= talent_solution_client.patch_project_company(company_name, update_company_request)
+		puts "Company updated: #{company_updated.to_json}"
+		return company_updated
 	rescue
 		puts "Got exception while updating company"
-		splitted_name = company_name.split('/');
+		splitted_name = company_name.split('/')
```
[the remark this answers is not in the document]
Comment (baochanzheng) [reply jobs/V3/basic_company_sample.rb]: Since the company name has to follow the format "projects/`project_id`/companies/`company_id`" and it's user input. The failure can be caused by bad-formatted company name. I am checking whether the user input is valid.

File: jobs/V3/basic_company_sample.rb [hunk 17]
```diff
@@ -12,191 +12,205 @@
...
 	jobs = Google::Apis::JobsV3
-	company_name = "companyName:"+ SecureRandom.hex;
-	company_generated = jobs::Company.new;
-	company_generated.display_name = "Google";
-	company_generated.headquarters_address= "1600 Amphitheatre Parkway Mountain View, CA 94043";
-	company_generated.external_id = company_name;
-	puts "Company generated: " + company_generated.to_json;
-	return company_generated;
-	# [END generate company]
+	company_name = "companyName: #{display_name}" + SecureRandom.hex
```
[the remark this answers is not in the document]
Comment (baochanzheng) [reply jobs/V3/basic_company_sample.rb]: Since external_id has to be unique, I am using SecureRandom.hex to generate the unique string. Is there a better way to do hat instead of using random hex?

File: jobs/V3/basic_job_sample.rb [hunk 18]
```diff
@@ -12,171 +12,172 @@
...
+
+	job_generated.compensation_info = compensation_info
+	puts "Job generated: #{job_generated.to_json}"
+	return job_generated
 	# [END generate basic job]
 end
 
-def create_job(job_to_be_created)
+def job_discovery_create_job job_to_be_created:, default_project_id:
```
[the remark this answers is not in the document]
Comment (baochanzheng) [reply jobs/V3/basic_job_sample.rb]: I didn't quite get it... you mean instead of passing project_id as a parm. I should get project id by parsing job_to_be_created?

File: jobs/spec/samples_spec.rb (hunk 10 above)
Comment (baochanzheng) [reply jobs/spec/samples_spec.rb]: Hi Frank, I don't quiet understand it. I checked Storage Ruby Sample which you recommended as example. It did the same thing as I did, verifying the output string. 
I used job service in specific test functions not in spec file. In test functions, I put error msgs like "Exception occurred..." I can check expect(capture).not_to include("Exception") to verify (indirectly) job service didn't go wrong. Is that what you want?
Can you show me an example of how to do that?
Thanks and regards, Baochan

Comment (frankyn) [reply jobs/spec/samples_spec.rb]: I think I missed this comment during Friday's review. 

I meant to verify the state of the service. If a job is created, check that it exists outside of the test.

Here's an example: https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/master/storage/spec/files_spec.rb#L231
In this example I request the Storage resource after the sample is executed to verify content-type was changed successfully.

Comment (baochanzheng) [reply jobs/spec/samples_spec.rb]: I got it! 
I am almost done changing all the files. I am doing multiple tasks simultaneously. Sorry for the delay.

**Comment by baochanzheng** [comment]
@frankyn Hi Frank, I applied the suggested changes to all the files. Pls take a look and let me know what is missing... (apologize ahead if I forgot something. There are too many changes. I am afraid there must be something I missed even though I did a really careful check before commit)

File: jobs/V3/auto_complete_sample.rb [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+											  project_id: default_project_id
+	when "default_auto_complete"
+		job_discovery_default_auto_complete company_name: company_name, 
+										    query: arguments.shift, 
+										    project_id: default_project_id
+	else
+	puts <<-usage
+Usage: bundle exec ruby auto_complete_sample.rb [command] [arguments]
+Commands:
+  job_title_auto_complete     <company_id><title_prefix>     Auto completes job titles within given company_name and title prefix
```
Comment (frankyn) [inline jobs/V3/auto_complete_sample.rb]: space between `<company_id>` and `<title_prefix>`

File: jobs/V3/basic_company_sample.rb [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+def job_discovery_generate_company display_name:, headquarters_address:
+	# [START generate_company]
+	# display_name         = "Your company display name"
+	# headquarters_address = "Your company headquarters address"
+
+	require "google/apis/jobs_v3"
+	require "securerandom"
+
+	jobs = Google::Apis::JobsV3
+	external_id = "externalId: #{display_name} #{SecureRandom.hex}"
```
Comment (frankyn) [inline jobs/V3/basic_company_sample.rb]: I still don't agree with `SecureRandom.hex`. I think this can be passed in with the `display_name`

File: jobs/V3/batch_operation_sample.rb [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+					             employment_types: ["FULL_TIME"],
+					             language_code: "en-US",
+					             application_info:
+					                (jobs::ApplicationInfo.new uris: ["http://careers.google.com"]),
+					             description: "System Administrator for software."
+
+  create_job_request1 = jobs::CreateJobRequest.new job: job_generated1
+  create_job_request2 = jobs::CreateJobRequest.new job: job_generated2
+
+  talent_solution_client.batch do |s|
```
Comment (frankyn) [inline jobs/V3/batch_operation_sample.rb]: what is `s`? Could you expand it to a full name such as `client`?

File: jobs/V3/batch_operation_sample.rb [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+  jobs_updated = Array.new
+  update_job_with_mask_requests = Array.new
+  job_to_be_updated.each{ |job|
+    request = jobs::UpdateJobRequest.new job: job,
+                       					 update_mask: "title"
+    update_job_with_mask_requests.push request
+  }
+
+  talent_solution_client.batch do |s|
+    update_job_with_mask_requests.each{ |update_job_with_mask_request|
```
Comment (frankyn) [inline jobs/V3/batch_operation_sample.rb]: Use `do..end` instead of `{..}`

File: jobs/V3/featured_job_sample.rb [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+
+  command = arguments.shift
+  default_project_id = "projects/#{ENV["GOOGLE_CLOUD_PROJECT"]}"
+  company_name = "#{default_project_id}/companies/#{arguments.shift}"
+
+  case command
+  when "featured_jobs_search"
+    company_got_test = job_discovery_get_company company_name: company_name
+    job_generated_test = job_discovery_generate_featured_job company_name: company_name
+    job_created_test = job_discovery_create_job job_to_be_created: job_generated_test,
```
Comment (frankyn) [inline jobs/V3/featured_job_sample.rb]: The prior three lines can be in the system tests for this sample instead of being included here. A user should be using the CLI to access an existing job set instead of creating a dummy job before they search.
```ruby
company_got_test = job_discovery_get_company company_name: company_name
job_generated_test = job_discovery_generate_featured_job company_name: company_name
job_created_test = job_discovery_create_job job_to_be_created: job_generated_test,
```

File: jobs/V3/auto_complete_sample.rb [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def job_discovery_job_title_auto_complete company_name:, query:, project_id:
+  # [START job_discovery_job_title_auto_complete]
+  # company_name  = "The company's name which the job belongs to"
+  # query         = "Job title prefix as auto complete query"
+  # project_id    = "Project id required"
```
Comment (frankyn) [inline jobs/V3/auto_complete_sample.rb]: Put `project_id` first and use "Google Cloud Project Id".

File: jobs/V3/auto_complete_sample.rb [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def job_discovery_job_title_auto_complete company_name:, query:, project_id:
+  # [START job_discovery_job_title_auto_complete]
+  # company_name  = "The company's name which the job belongs to"
+  # query         = "Job title prefix as auto complete query"
```
Comment (frankyn) [inline jobs/V3/auto_complete_sample.rb]: Use the wording from the API reference docs for attributes. 
For example `query`: The query string that matches against the job title, description, and location fields.

https://cloud.google.com/talent-solution/job-search/docs/reference/rest/v3/JobQuery

**Comment by xinyunh0929** [comment]
Hi Ruby-reviewers, we are on the way to the gRPC client library (Veneer). 
However, to give our current customer who is using discovery-based library (Apiary) a better experience, we want to provide more code sample in different languages, and that's the reason we have this pull request.
For the new code samples within gRPC, we have already contacted Justin and Yoshi team, and I believe that they are coming soon.

**Comment by frankyn** [comment]
@baochanzheng I want one more pass from @GoogleCloudPlatform/yoshi-ruby, because  I'll be OOO next week. I don't have bandwidth to review it today apologies and I also don't want to leave you waiting.

Summoning: @GoogleCloudPlatform/yoshi-ruby could someone give this one last review. I think it should be close to done, but didn't have time to do one more pass. Thank you in advance!

**Comment by baochanzheng** [comment]
Hi Frank, I hope you had a wonderful holiday season!

I just noticed the Kokoro CL failure in this pull request. Since they always successed in my local testing,  I suspect it's just timing issue. I tried to fix this issue by adding sleep time to those two failed test, but I don't know how to force run kokoro test to verify my fix. Can you show me how?

Thanks and best,
Baochan

**Comment by baochanzheng** [comment]
@frankyn Hi Frank, have you been back from vacation yet? Could you review this pull request again? I'd really like to get things done before I forgot everything... It's been taking too long....
Let me know your feedback, especially how the Kokoro CL failure issue goes. Thanks!

Best,
Baochan

**Comment by frankyn** [comment]
Started the Kokoro jobs. I'm got back this week and slowly ramping back up. Thank you for your patience! Looking at this now.

**Comment by baochanzheng** [comment]
@frankyn 
Can you rerun the test again? We gathered as a team to discuss about this issue... It's weird that the test passed in my local run but failed in your presubmit run. We suspect the time for indexing phase differs case by case, so we might need to wait longer before we check the result...
BTW, is there a way that I can run the test myself without bothering you? I feel so bad asking you to add the tag again and again :(
